lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -0,0 +1,893 @@
+
+          Operation *termOp = subModuleOp.getBody().front().getTerminator();
+          Location insertLoc = termOp->getLoc();
+          rewriter.setInsertionPoint(termOp);
+
+          ValueVectorList portList =
+              extractSubfields(subModuleOp, insertLoc, rewriter);
+
+          if (isa<mlir::AddIOp>(op))
+            buildBinaryLogic<AddPrimOp>(portList, insertLoc, rewriter);
+          else if (isa<mlir::SubIOp>(op))
+            buildBinaryLogic<SubPrimOp>(portList, insertLoc, rewriter);
+          else if (isa<mlir::MulIOp>(op))
+            buildBinaryLogic<MulPrimOp>(portList, insertLoc, rewriter);
+
+          else if (isa<mlir::AndOp>(op))
+            buildBinaryLogic<AndPrimOp>(portList, insertLoc, rewriter);
+          else if (isa<mlir::OrOp>(op))
+            buildBinaryLogic<OrPrimOp>(portList, insertLoc, rewriter);
+          else if (isa<mlir::XOrOp>(op))
+            buildBinaryLogic<XorPrimOp>(portList, insertLoc, rewriter);
+
+          else if (auto cmpOp = dyn_cast<mlir::CmpIOp>(op)) {
+            auto cmpOpAttr = cmpOp.getPredicate();
+            switch (cmpOpAttr) {
+            case CmpIPredicate::eq:
+              buildBinaryLogic<EQPrimOp>(portList, insertLoc, rewriter);
+              break;
+            case CmpIPredicate::ne:
+              buildBinaryLogic<NEQPrimOp>(portList, insertLoc, rewriter);
+              break;
+            case CmpIPredicate::slt:
+              buildBinaryLogic<LTPrimOp>(portList, insertLoc, rewriter);
+              break;
+            case CmpIPredicate::sle:
+              buildBinaryLogic<LEQPrimOp>(portList, insertLoc, rewriter);
+              break;
+            case CmpIPredicate::sgt:
+              buildBinaryLogic<GTPrimOp>(portList, insertLoc, rewriter);
+              break;
+            case CmpIPredicate::sge:
+              buildBinaryLogic<GEQPrimOp>(portList, insertLoc, rewriter);
+              break;
+            }
+          } else if (isa<mlir::ShiftLeftOp>(op))
+            buildBinaryLogic<DShlPrimOp>(portList, insertLoc, rewriter);
+          else if (isa<mlir::SignedShiftRightOp>(op))
+            buildBinaryLogic<DShrPrimOp>(portList, insertLoc, rewriter);
+
+          // Build elastic components logic
+          else if (isa<handshake::MergeOp>(op))
+            buildMergeLogic(portList, insertLoc, rewriter);
+          else if (isa<handshake::ControlMergeOp>(op))
+            buildControlMergeLogic(op, portList, insertLoc, rewriter);
+
+          else if (isa<handshake::MuxOp>(op))
+            buildMuxLogic(portList, insertLoc, rewriter);
+
+          else if (isa<handshake::BranchOp>(op))
+            buildBranchLogic(op, portList, insertLoc, rewriter);
+          else if (isa<handshake::ConditionalBranchOp>(op))
+            buildConditionalBranchLogic(op, portList, insertLoc, rewriter);
+
+          else if (isa<handshake::ForkOp>(op))
+            buildForkLogic(op, portList, insertLoc, rewriter);
```

Comment:
These should all use 'dyn_cast' and pass a typed operation into the callee instead of passing operation&

Reply:
In a followup patch it might be worth investigating whether a visitor for handshake would be a good thing to invest in, similar to the firrtl/rtl ones.